File: plugins/statusline/scripts/statusline.py
#!/usr/bin/env python3
"""Three-line Claude Code statusline for NerdFont, true-colour terminals.

Reads the statusline JSON on stdin and prints:
  1. vim mode, git repo and path, branch, lines added and removed
  2. model, effort, output style, context window usage
  3. rate-limit reset, cost, session usage, weekly usage
"""

from __future__ import annotations

import json
import os
import re
import subprocess
import sys
from dataclasses import dataclass
from datetime import datetime, timezone

RESET = "\033[0m"
BOLD = "\033[1m"
DIM = "\033[2m"


def rgb(r: int, g: int, b: int) -> str:
    return f"\033[38;2;{r};{g};{b}m"


@dataclass(frozen=True)
class Palette:
    lavender: str
    sky: str
    mint: str
    gold: str
    coral: str
    rose: str
    silver: str
    muted: str
    bar_ok: str
    bar_warn: str
    bar_crit: str
    vim_normal: str
    vim_insert: str
    vim_visual: str
    vim_command: str
    effort_low: str
    effort_max: str


# Catppuccin Macchiato
DARK = Palette(
    lavender=rgb(183, 189, 248),
    sky=rgb(145, 215, 227),
    mint=rgb(166, 218, 149),
    gold=rgb(238, 212, 159),
    coral=rgb(245, 169, 127),
    rose=rgb(238, 153, 160),
    silver=rgb(165, 173, 203),
    muted=rgb(110, 115, 141),
    bar_ok=rgb(139, 213, 202),
    bar_warn=rgb(238, 212, 159),
    bar_crit=rgb(237, 135, 150),
    vim_normal=rgb(138, 173, 244),
    vim_insert=rgb(166, 218, 149),
    vim_visual=rgb(198, 160, 246),
    vim_command=rgb(237, 135, 150),
    effort_low=rgb(125, 196, 228),
    effort_max=rgb(237, 135, 150),
)

# Catppuccin Latte
LIGHT = Palette(
    lavender=rgb(114, 135, 253),
    sky=rgb(4, 165, 229),
    mint=rgb(64, 160, 43),
    gold=rgb(223, 142, 29),
    coral=rgb(254, 100, 11),
    rose=rgb(230, 69, 83),
    silver=rgb(108, 111, 133),
    muted=rgb(140, 143, 161),
    bar_ok=rgb(23, 146, 153),
    bar_warn=rgb(223, 142, 29),
    bar_crit=rgb(210, 15, 57),
    vim_normal=rgb(30, 102, 245),
    vim_insert=rgb(64, 160, 43),
    vim_visual=rgb(136, 57, 239),
    vim_command=rgb(210, 15, 57),
    effort_low=rgb(32, 159, 181),
    effort_max=rgb(210, 15, 57),
)

PALETTES = {"dark": DARK, "light": LIGHT}

ICO_MODEL = "\uf135"
ICO_BOLT = "\uf0e7"
ICO_DB = "\ue64d"
ICO_CLOCK = "\U000f13ab"
ICO_DOLLAR = "\uf252"
ICO_CAL = "\U000f1a32"
ICO_VIM = "\ue62b"
ICO_GIT = "\ueafc"
ICO_BRANCH = "\ue725"
ICO_STYLE = "\U000f150f"

EM_DASH = "\u2014"

CIRCLES = ("\u25ef", "\u25d4", "\u25d1", "\u25d5", "\u25cf")

EFFORTS = {
    "low": ("\u2581 low", "effort_low"),
    "medium": ("\u2582 med", "gold"),
    "high": ("\u2583 high", "coral"),
    "xhigh": ("\u2585 xhigh", "rose"),
    "max": ("\u2588 max", "effort_max"),
}

VIM_MODES = {
    "NORMAL": "vim_normal",
    "INSERT": "vim_insert",
    "VISUAL": "vim_visual",
    "V-LINE": "vim_visual",
    "V-BLOCK": "vim_visual",
    "COMMAND": "vim_command",
    "REPLACE": "coral",
}

MODEL_PREFIXES = (("Claude Opus", "Opus"), ("Claude Sonnet", "Sonnet"), ("Claude Haiku", "Haiku"))

# Skip huge untracked files when counting added lines; they are almost never source.
MAX_UNTRACKED_BYTES = 10 * 1024 * 1024


def trigger_file() -> str:
    return os.path.expanduser("~/.local/share/theme-monitor/theme-change.trigger")


def macos_appearance() -> str | None:
    if sys.platform != "darwin":
        return None
    try:
        result = subprocess.run(
            ["defaults", "read", "-g", "AppleInterfaceStyle"],
            capture_output=True,
            text=True,
            timeout=1,
        )
    except (OSError, subprocess.SubprocessError):
        return None
    # The key only exists in dark mode, so a failed read means light.
    return "dark" if result.returncode == 0 and result.stdout.strip() == "Dark" else "light"


def detect_theme() -> str:
    override = os.environ.get("CLAUDE_STATUSLINE_THEME", "").strip().lower()
    if override in PALETTES:
        return override
    try:
        with open(trigger_file()) as f:
            value = f.read().strip().lower()
        if value in PALETTES:
            return value
    except OSError:
        pass
    return macos_appearance() or "dark"


def get(d: object, *keys: str) -> object:
    for key in keys:
        if not isinstance(d, dict):
            return None
        d = d.get(key)
    return d


def number(value: object) -> float | None:
    if isinstance(value, bool):
        return None
    try:
        return float(value)  # type: ignore[arg-type]
    except (TypeError, ValueError):
        return None


def pct_color(p: Palette, pct: float) -> str:
    return p.bar_ok if pct < 60 else (p.bar_warn if pct < 85 else p.bar_crit)


def progress_circle(p: Palette, pct: float) -> str:
    idx = max(0, min(len(CIRCLES) - 1, round(pct / 100 * (len(CIRCLES) - 1))))
    return f"{pct_color(p, pct)}{CIRCLES[idx]}{RESET}"


def fmt_tokens(n: int) -> str:
    if n >= 1_000_000:
        return f"{n / 1_000_000:.1f}M"
    if n >= 1_000:
        return f"{n / 1_000:.0f}k"
    return str(n)


def fmt_cost(usd: float) -> str:
    if usd < 0.001:
        return "$0"
    if usd < 1.0:
        return f"${usd:.3f}"
    return f"${usd:.2f}"


def fmt_duration(seconds: int) -> str:
    h, rest = divmod(seconds, 3600)
    m, s = divmod(rest, 60)
    return f"{h}h {m:02d}m" if h else f"{m}m {s:02d}s"


def fmt_resets(value: object, now: datetime | None = None) -> str:
    try:
        if isinstance(value, (int, float)):
            seconds = value / 1000 if value >= 1e10 else value
            dt = datetime.fromtimestamp(seconds, tz=timezone.utc)
        else:
            dt = datetime.fromisoformat(str(value).replace("Z", "+00:00"))
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        sec = int((dt - (now or datetime.now(timezone.utc))).total_seconds())
    except (TypeError, ValueError, OverflowError, OSError):
        return "?"
    if sec <= 0:
        return "now"
    h, rest = divmod(sec, 3600)
    m = rest // 60
    return f"{h}h {m:02d}m" if h else f"{m}m"


def _git(cwd: str, *args: str) -> str:
    try:
        result = subprocess.run(
            ["git", *args],
            cwd=cwd,
            capture_output=True,
            encoding="utf-8",
            errors="replace",
            timeout=2,
            # Refreshing every second must not contend for the index lock.
            env={**os.environ, "GIT_OPTIONAL_LOCKS": "0"},
        )
    except (OSError, subprocess.SubprocessError):
        return ""
    return result.stdout.strip() if result.returncode == 0 else ""


def _count_lines(root: str, names: list[str]) -> int:
    total = 0
    for name in names:
        path = os.path.join(root, name)
        try:
            if not os.path.isfile(path) or os.path.getsize(path) > MAX_UNTRACKED_BYTES:
                continue
            with open(path, "rb") as f:
                total += f.read().count(b"\n")
        except OSError:
            continue
    return total


def git_info(cwd: str) -> dict | None:
    root = _git(cwd, "rev-parse", "--show-toplevel")
    if not root:
        return None
    branch = _git(cwd, "branch", "--show-current") or _git(cwd, "rev-parse", "--short", "HEAD")
    stat = _git(root, "diff", "--shortstat", "HEAD")
    added = re.search(r"(\d+) insertion", stat)
    deleted = re.search(r"(\d+) deletion", stat)
    untracked = _git(root, "ls-files", "-z", "--others", "--exclude-standard")
    return {
        "repo": os.path.basename(root),
        "branch": branch,
        "rel": _git(cwd, "rev-parse", "--show-prefix").rstrip("/"),
        "added": (int(added.group(1)) if added else 0)
        + _count_lines(root, [n for n in untracked.split("\0") if n]),
        "deleted": int(deleted.group(1)) if deleted else 0,
    }


def sep(p: Palette) -> str:
    return f"  {p.muted}\u00b7{RESET}  "


def vim_badge(p: Palette, mode: str) -> str:
    m = mode.upper()
    col = getattr(p, VIM_MODES.get(m, "silver"))
    return f"{BOLD}{col}{ICO_VIM} {m}{RESET}"


def effort_badge(p: Palette, level: str) -> str:
    label, attr = EFFORTS.get(level, (level, "silver"))
    return f"{BOLD}{getattr(p, attr)}{ICO_BOLT} {label}{RESET}"


def line1(d: dict, p: Palette) -> str:
    cwd = get(d, "workspace", "current_dir") or get(d, "cwd") or os.getcwd()
    parts: list[str] = []

    mode = get(d, "vim", "mode")
    if isinstance(mode, str) and mode:
        parts.append(vim_badge(p, mode))

    gi = git_info(str(cwd))
    if gi:
        repo = f"{BOLD}{p.lavender}{gi['repo']}{RESET}"
        if gi["rel"]:
            repo += f"{p.muted}/{gi['rel']}{RESET}"
        parts.append(f"{BOLD}{p.lavender}{ICO_GIT} {RESET}{repo}")
        if gi["branch"]:
            parts.append(f"{p.gold}{ICO_BRANCH} {gi['branch']}{RESET}")
        changes = []
        if gi["added"]:
            changes.append(f"{p.mint}+{gi['added']}{RESET}")
        if gi["deleted"]:
            changes.append(f"{p.bar_crit}-{gi['deleted']}{RESET}")
        if changes:
            parts.append(" ".join(changes))
    else:
        path = str(cwd)
        home = os.path.expanduser("~")
        if path == home or path.startswith(home + os.sep):
            path = "~" + path[len(home) :]
        parts.append(f"{p.silver}{path}{RESET}")

    return sep(p).join(parts)


def line2(d: dict, p: Palette) -> str:
    name = get(d, "model", "display_name")
    name = name if isinstance(name, str) else ""
    for long, short in MODEL_PREFIXES:
        name = name.replace(long, short)
    model = f"{BOLD}{p.lavender}{ICO_MODEL} {name or EM_DASH}{RESET}"

    inner = []
    level = get(d, "effort", "level")
    if isinstance(level, str) and level:
        inner.append(effort_badge(p, level))
    style = get(d, "output_style", "name")
    if isinstance(style, str) and style and style.lower() != "default":
        inner.append(f"{p.sky}{ICO_STYLE} {style}{RESET}")
    modes = f"{p.muted} | {RESET}".join(inner)

    pct = number(get(d, "context_window", "used_percentage")) or 0.0
    total = int(number(get(d, "context_window", "context_window_size")) or 0)
    col = pct_color(p, pct)
    ctx = f"{col}{ICO_DB}{RESET} {progress_circle(p, pct)} {BOLD}{col}{pct:.0f}%{RESET}"
    if total:
        used = int(pct / 100 * total)
        ctx += f"  {DIM}({fmt_tokens(used)}/{fmt_tokens(total)}){RESET}"

    return sep(p).join(x for x in (model, modes, ctx) if x)


def line3(d: dict, p: Palette) -> str:
    parts: list[str] = []

    resets = get(d, "rate_limits", "five_hour", "resets_at")
    if resets:
        parts.append(f"{p.sky}{ICO_CLOCK} {fmt_resets(resets)}{RESET}")

    usd = number(get(d, "cost", "total_cost_usd"))
    session = number(get(d, "rate_limits", "five_hour", "used_percentage"))
    if usd is not None and usd > 0:
        s = f"{p.mint}{ICO_DOLLAR} {fmt_cost(usd)}{RESET}"
        if session is not None:
            s += f"  {pct_color(p, session)}{session:.0f}% session{RESET}"
        parts.append(s)
    elif session is not None:
        parts.append(f"{pct_color(p, session)}{ICO_DOLLAR} {session:.0f}% session{RESET}")

    weekly = number(get(d, "rate_limits", "seven_day", "used_percentage"))
    if weekly is not None:
        parts.append(f"{pct_color(p, weekly)}{ICO_CAL}  {weekly:.0f}% weekly{RESET}")

    if not parts:
        ms = number(get(d, "cost", "total_duration_ms"))
        if ms:
            parts.append(f"{p.sky}{ICO_CLOCK} {fmt_duration(int(ms / 1000))}{RESET}")

    return sep(p).join(parts)


def render(data: dict, theme: str) -> str:
    p = PALETTES[theme]
    out = []
    for fn in (line1, line2, line3):
        try:
            line = fn(data, p)
        except Exception as exc:
            line = f"{p.bar_crit}statusline error: {exc}{RESET}"
        if line:
            out.append(line)
    return "\n".join(out) + "\n"


def main() -> None:
    try:
        data = json.loads(sys.stdin.read() or "{}")
    except ValueError:
        data = {}
    if not isinstance(data, dict):
        data = {}
    sys.stdout.write(render(data, detect_theme()))


if __name__ == "__main__":
    main()

File: pi/extensions/statusline.ts
import { execFileSync } from "node:child_process";
import { readFileSync } from "node:fs";
import { homedir } from "node:os";
import { basename, relative, resolve } from "node:path";
import type { AssistantMessage } from "@earendil-works/pi-ai";
import type {
	ExtensionAPI,
	ExtensionContext,
} from "@earendil-works/pi-coding-agent";
import { truncateToWidth } from "@earendil-works/pi-tui";

type FooterFactory = Parameters<
	NonNullable<ExtensionContext["ui"]["setFooter"]>
>[0];

type GitInfo = {
	repo: string;
	branch: string;
	rel: string;
	added: number;
	deleted: number;
};

const ICON = {
	model: "",
	bolt: "",
	db: "",
	clock: "󱎫",
	dollar: "",
	calendar: "󱨲",
	git: "",
	branch: "",
	style: "󱔏",
};

const SEP = "  ·  ";
const RESET = "\x1b[0m";
const BOLD = "\x1b[1m";

type Palette = {
	lavender: string;
	sky: string;
	mint: string;
	gold: string;
	coral: string;
	rose: string;
	silver: string;
	muted: string;
	barOk: string;
	barWarn: string;
	barCrit: string;
	effortLow: string;
	effortMax: string;
};

function rgb(red: number, green: number, blue: number): string {
	return `\x1b[38;2;${red};${green};${blue}m`;
}

const DARK: Palette = {
	lavender: rgb(186, 187, 241),
	sky: rgb(153, 209, 219),
	mint: rgb(166, 209, 137),
	gold: rgb(229, 200, 144),
	coral: rgb(239, 159, 118),
	rose: rgb(234, 153, 156),
	silver: rgb(165, 173, 206),
	muted: rgb(115, 121, 148),
	barOk: rgb(129, 200, 190),
	barWarn: rgb(229, 200, 144),
	barCrit: rgb(231, 130, 132),
	effortLow: rgb(133, 193, 220),
	effortMax: rgb(231, 130, 132),
};

const LIGHT: Palette = {
	lavender: rgb(114, 135, 253),
	sky: rgb(4, 165, 229),
	mint: rgb(64, 160, 43),
	gold: rgb(223, 142, 29),
	coral: rgb(254, 100, 11),
	rose: rgb(230, 69, 83),
	silver: rgb(108, 111, 133),
	muted: rgb(140, 143, 161),
	barOk: rgb(23, 146, 153),
	barWarn: rgb(223, 142, 29),
	barCrit: rgb(210, 15, 57),
	effortLow: rgb(32, 159, 181),
	effortMax: rgb(210, 15, 57),
};

function palette(): Palette {
	try {
		return readFileSync(
			resolve(homedir(), ".local/share/theme-monitor/theme-change.trigger"),
			"utf8",
		).trim() === "light"
			? LIGHT
			: DARK;
	} catch {
		return DARK;
	}
}

function tildify(path: string): string {
	const home = homedir();
	return path === home || path.startsWith(`${home}/`)
		? `~${path.slice(home.length)}`
		: path;
}

function paint(color: string, text: string, bold = false): string {
	return `${bold ? BOLD : ""}${color}${text}${RESET}`;
}

function command(cwd: string, args: string[]): string {
	try {
		return execFileSync("git", args, {
			cwd,
			encoding: "utf8",
			stdio: ["ignore", "pipe", "ignore"],
			timeout: 500,
		}).trim();
	} catch {
		return "";
	}
}

function gitInfo(cwd: string): GitInfo | undefined {
	const root = command(cwd, ["rev-parse", "--show-toplevel"]);
	if (!root) return undefined;

	const stat = command(cwd, ["diff", "--shortstat", "HEAD"]);
	const added = Number(stat.match(/(\d+) insertion/)?.[1] ?? 0);
	const deleted = Number(stat.match(/(\d+) deletion/)?.[1] ?? 0);
	const untracked = command(cwd, [
		"ls-files",
		"--others",
		"--exclude-standard",
	]);
	let untrackedLines = 0;
	for (const file of untracked.split("\n").filter(Boolean)) {
		try {
			untrackedLines +=
				readFileSync(resolve(root, file), "utf8").split("\n").length - 1;
		} catch {
			// The file may disappear while the footer is rendering.
		}
	}

	const rel = relative(root, cwd);
	return {
		repo: basename(root),
		branch:
			command(cwd, ["branch", "--show-current"]) ||
			command(cwd, ["rev-parse", "--short", "HEAD"]),
		rel: rel === "." ? "" : rel,
		added: added + untrackedLines,
		deleted,
	};
}

function formatTokens(value: number): string {
	if (value >= 1_000_000) return `${(value / 1_000_000).toFixed(1)}M`;
	if (value >= 1_000) return `${Math.round(value / 1_000)}k`;
	return String(value);
}

function formatCost(value: number): string {
	if (value < 0.001) return "$0";
	if (value < 1) return `$${value.toFixed(3)}`;
	return `$${value.toFixed(2)}`;
}

function effort(level: string): string {
	return (
		(
			{
				low: "▁ low",
				medium: "▂ med",
				high: "▃ high",
				xhigh: "▅ xhigh",
				max: "█ max",
			} as Record<string, string>
		)[level] ?? level
	);
}

function effortColor(level: string, colors: Palette): string {
	return (
		(
			{
				low: colors.effortLow,
				medium: colors.gold,
				high: colors.coral,
				xhigh: colors.rose,
				max: colors.effortMax,
			} as Record<string, string>
		)[level] ?? colors.silver
	);
}

function usage(ctx: ExtensionContext) {
	const value = ctx.getContextUsage();
	const percent = value?.percent ?? 0;
	const window = value?.contextWindow ?? ctx.model?.contextWindow ?? 0;
	return {
		percent,
		window,
		tokens: value?.tokens ?? Math.round((percent / 100) * window),
	};
}

function sessionTotals(ctx: ExtensionContext) {
	let input = 0;
	let output = 0;
	let cost = 0;
	for (const entry of ctx.sessionManager.getBranch()) {
		if (entry.type !== "message" || entry.message.role !== "assistant")
			continue;
		const message = entry.message as AssistantMessage;
		input += message.usage.input;
		output += message.usage.output;
		cost += message.usage.cost.total;
	}
	return { input, output, cost };
}

export default function (pi: ExtensionAPI) {
	let startupTimer: NodeJS.Timeout | undefined;
	let installTimer: NodeJS.Timeout | undefined;

	pi.on("session_start", (_event, ctx) => {
		let tui: { requestRender(): void } | undefined;

		const footer: FooterFactory = (_tui, theme, footerData) => {
			tui = _tui;
			const unsubscribe = footerData.onBranchChange(() => tui?.requestRender());

			return {
				dispose: unsubscribe,
				invalidate() {},
				render(width: number): string[] {
					const colors = palette();
					const separator = `  ${colors.muted}·${RESET}  `;
					const cwd = ctx.cwd;
					const git = gitInfo(cwd);
					const line1Parts: string[] = [];
					if (git) {
						let repoPart = paint(
							colors.lavender,
							`${ICON.git} ${git.repo}`,
							true,
						);
						if (git.rel) repoPart += paint(colors.muted, `/${git.rel}`);
						line1Parts.push(repoPart);
						if (git.branch)
							line1Parts.push(
								paint(colors.gold, `${ICON.branch} ${git.branch}`),
							);
						if (git.added) line1Parts.push(paint(colors.mint, `+${git.added}`));
						if (git.deleted)
							line1Parts.push(paint(colors.barCrit, `-${git.deleted}`));
					} else {
						line1Parts.push(
							paint(colors.silver, tildify(cwd)),
						);
					}

					const model = ctx.model?.id ?? "no-model";
					const level = ctx.thinkingLevel ?? "off";
					const context = usage(ctx);
					let bar = "○";
					let contextColor: "success" | "warning" | "error" = "success";
					if (context.percent >= 85) {
						bar = "●";
						contextColor = "error";
					} else if (context.percent >= 60) {
						bar = "◕";
						contextColor = "warning";
					}
					let contextBarColor = colors.barOk;
					if (contextColor === "warning") contextBarColor = colors.barWarn;
					if (contextColor === "error") contextBarColor = colors.barCrit;
					const contextTotal =
						context.tokens == null
							? `?/${formatTokens(context.window)}`
							: `(${formatTokens(context.tokens)}/${formatTokens(context.window)})`;
					const contextPart =
						paint(contextBarColor, `${ICON.db} ${bar} `) +
						paint(contextBarColor, `${context.percent.toFixed(0)}%`, true) +
						`  ${colors.muted}${contextTotal}${RESET}`;
					const line2Parts = [
						paint(colors.lavender, `${ICON.model} ${model}`, true),
						level !== "off"
							? paint(
									effortColor(level, colors),
									`${ICON.bolt} ${effort(level)}`,
									true,
								)
							: "",
						contextPart,
					].filter(Boolean);

					const totals = sessionTotals(ctx);
					const line3Parts = [
						paint(
							colors.sky,
							`${ICON.clock} ${formatTokens(totals.input)}↑ ${formatTokens(totals.output)}↓`,
						),
						paint(colors.mint, `${ICON.dollar} ${formatCost(totals.cost)}`),
						paint(colors.sky, `${ICON.calendar} session`),
					];

					const lines = [
						line1Parts.join(separator),
						line2Parts.join(separator),
						line3Parts.join(separator),
					];
					return lines.map((line) => truncateToWidth(line, width, ""));
				},
			};
		};

		const installFooter = () => {
			try {
				ctx.ui.setFooter(footer);
				tui?.requestRender();
			} catch {
				// Pi invalidates the old context during /reload or session replacement.
			}
		};

		// pi-claude-code-tui restores Pi's default footer from its session_start
		// handler. Install ours both immediately and after startup handlers finish.
		installFooter();
		if (startupTimer) clearInterval(startupTimer);
		startupTimer = setInterval(installFooter, 250);
		installTimer = setTimeout(() => {
			if (startupTimer) clearInterval(startupTimer);
			startupTimer = undefined;
		}, 10_000);

		pi.registerCommand("statusline", {
			description: "Install the custom statusline footer",
			handler: async (_args, commandCtx) => {
				commandCtx.ui.setFooter(footer);
				tui?.requestRender();
				commandCtx.ui.notify("Custom statusline enabled", "info");
			},
		});

		const refresh = () => {
			installFooter();
			tui?.requestRender();
		};
		pi.on("message_update", refresh);
		pi.on("message_end", refresh);
		pi.on("model_select", refresh);
		pi.on("thinking_level_select", refresh);
		pi.on("turn_end", refresh);
		pi.on("agent_start", refresh);
	});

	pi.on("session_shutdown", () => {
		if (startupTimer) clearInterval(startupTimer);
		if (installTimer) clearTimeout(installTimer);
		startupTimer = undefined;
		installTimer = undefined;
	});
}
